doc/pitfalls/data_leakage.rst

```diff
@@ -0,0 +1,152 @@
+improve your model build time and performance. Feature selection should
+**only** use the training data. Including the test data in feature selection
+will optimistically bias your model.
+
+To demonstrate we will create a binary classification problem with
+1000 randomly generated features::
+
+    >>> from numpy.random import default_rng
+    >>> n_samples, n_features, n_classes = 200, 10000, 2
+    >>> rng = default_rng(42)
+    >>> X = rng.standard_normal((n_samples, n_features))
+    >>> y = rng.choice(n_classes, n_samples)
+
+**Wrong**
+
+Using all the data to perform feature selection results in an accuracy score
+much higher than chance. Since our `X` and `y` are independent, we expect
+performance to be around 0.5. However, since the feature selection step
+'sees' the test data, the model has an unfair advantange.
+
+    >>> from sklearn.feature_selection import SelectKBest
+    >>> from sklearn.model_selection import cross_val_score
+    >>> from sklearn.ensemble import GradientBoostingClassifier
+    >>> X_selected = SelectKBest(k=25).fit_transform(X, y)
+    >>> scores = cross_val_score(GradientBoostingClassifier(random_state=1),
+    ...                          X_selected, y)
+    >>> print(f"Mean cross-validation accuracy: {scores.mean()}")
+    Mean cross-validation accuracy: 0.775
+
+**Right**
+
+To ensure that feature selection is performed using only the train dataset
+we will use a :class:`~sklearn.pipeline.Pipeline` to chain together the
+feature selection and model. Feeding our pipeline into
+:func:`~sklearn.model_selection.cross_val_score` ensures that only the
+training data is used for feature selection (and when fitting our model).
+The test data will only be used for calculating the accuracy score. The
+score is now what we would expect for the data, close to chance::
+
+    >>> from sklearn.pipeline import make_pipeline
+    >>> pipeline = make_pipeline(SelectKBest(k=25),
+    ...                          GradientBoostingClassifier(random_state=1))
+    >>> scores = cross_val_score(pipeline, X, y)
+    >>> print(f"Mean Accuracy: {scores.mean()}")
+    Mean Accuracy: 0.495
+
+Data leakage during imputation
+==============================
+
+There are a number of methods to impute missing values in data. For example,
+:class:`~sklearn.impute.SimpleImputer` allows you to replace missing values
+with the mean of that feature. Only the train data should be used to
+calculate this mean value as including the test data in the mean calculation
+will introduce information about the test data into the model.
+
+To demonstrate this, we will use the :ref:`diabetes_dataset` and
+artificially introduce missing values. We introduce a large number of
+missing values to make the effect more pronounced::
+
+    >>> import numpy as np
+    >>> from sklearn.datasets import load_diabetes
+    >>> X, y = load_diabetes(return_X_y=True)
+    >>> rng = np.random.RandomState(42)
```

Comment:
don't use want to use

from numpy.random import default_rng

as above?

Reply:
I'd rather not, we have zero use of `default_rng` throughout the codebase.

Also wouldn't this make results non-reproducible (and CI would complain)?

Reply:
BTW this fails anyway: default_rng returns a Generator, and we don't support them.

```py
In [1]: from numpy.random import default_rng                                                                           

In [2]: from sklearn.datasets import make_classification                                                               
make_
In [3]: make_classification(random_state=default_rng)

ValueError: Generator(PCG64) at 0x7F4B13886C70 cannot be used to seed a numpy.random.RandomState instance
```

Reply:
Changed to use `np.random.RandomState` as in other examples.

Reply:
another reason is that `default_rng` requires numpy 1.17 or above, scikit-learn only requires numpy 1.13